pyrealm/splash/evap.py

```diff
@@ -0,0 +1,316 @@
+    (kg/m^3) following :cite:t:`chen:2008a`.
+
+    Args:
+        tc: Air temperature (°C)
+        p: Atmospheric pressure (Pa)
+
+    Returns:
+        The calculated density of water
+    """
+
+    # TODO - merge
+
+    # Calculate density at 1 atm (kg/m^3):
+    po = 0.99983952 + (6.788260e-5) * tc
+    po += -(9.08659e-6) * tc * tc
+    po += (1.022130e-7) * tc * tc * tc
+    po += -(1.35439e-9) * tc * tc * tc * tc
+    po += (1.471150e-11) * tc * tc * tc * tc * tc
+    po += -(1.11663e-13) * tc * tc * tc * tc * tc * tc
+    po += (5.044070e-16) * tc * tc * tc * tc * tc * tc * tc
+    po += -(1.00659e-18) * tc * tc * tc * tc * tc * tc * tc * tc
+
+    # Calculate bulk modulus at 1 atm (bar):
+    ko = 19652.17 + 148.1830 * tc
+    ko += -2.29995 * tc * tc
+    ko += 0.01281 * tc * tc * tc
+    ko += -(4.91564e-5) * tc * tc * tc * tc
+    ko += (1.035530e-7) * tc * tc * tc * tc * tc
+
+    # Calculate temperature dependent coefficients:
+    ca = 3.26138 + (5.223e-4) * tc
+    ca += (1.324e-4) * tc * tc
+    ca += -(7.655e-7) * tc * tc * tc
+    ca += (8.584e-10) * tc * tc * tc * tc
+
+    cb = 7.2061e-5 + -(5.8948e-6) * tc
+    cb += (8.69900e-8) * tc * tc
+    cb += -(1.0100e-9) * tc * tc * tc
+    cb += (4.3220e-12) * tc * tc * tc * tc
+
+    # Convert atmospheric pressure to bar (1 bar = 100000 Pa)
+    pbar = (1.0e-5) * p
+
+    pw = ko + ca * pbar + cb * pbar**2.0
+    pw /= ko + ca * pbar + cb * pbar**2.0 - pbar
+    pw *= (1e3) * po
+    return pw
+
+
+def psychro(tc: np.ndarray, p: np.ndarray) -> np.ndarray:
+    r"""Calculate the psychrometric constant.
+
+    Calculates the psychrometric constant (:math:`\lambda`, Pa/K) given the temperature
+    and atmospheric pressure following :cite:t:`allen:1998a` and
+    :cite:t:`tsilingiris:2008a`.
+
+    Args:
+        tc: Air temperature (°C)
+        p: Atmospheric pressure (Pa)
+
+    Returns:
+        The calculated psychrometric constant
+    """
+
+    # Calculate the specific heat capacity of water, J/kg/K
+    cp = specific_heat(tc)
+
+    # Calculate latent heat of vaporization, J/kg
+    lv = enthalpy_vap(tc)
+
+    # Calculate psychrometric constant, Pa/K
+    # Eq. 8, Allen et al. (1998)
+    return cp * kMa * p / (kMv * lv)
+
+
+def specific_heat(tc: np.ndarray) -> np.ndarray:
+    """Calculate the specific heat of air.
+
+    Calculates the specific heat of air at a constant pressure (:math:`c_{pm}`, J/kg/K)
+    following :cite:t:`tsilingiris:2008a`. This equation is only valid for temperatures
+    between 0 and 100 °C.
+
+    Args:
+        tc: Air temperature (°C)
+
+    Returns:
+        The specific heat of air values.
+    """
+    tc = np.clip(tc, 0, 100)
+    cp = 1.0045714270 + (2.050632750e-3) * tc
+    cp += -(1.631537093e-4) * tc * tc
+    cp += (6.212300300e-6) * tc * tc * tc
+    cp += -(8.830478888e-8) * tc * tc * tc * tc
+    cp += (5.071307038e-10) * tc * tc * tc * tc * tc
+    cp *= 1e3
+
+    return cp
```

Comment:
I have a feeling that there must be simpler approaches to calculate specific heat of air and water density?

Reply:
Perhaps this would be simpler: 
```
p = np.poly1d([1.0045714270, 2.050632750e-3, -1.631537093e-4, ...])
cp = p(tc) * 1e3
```

Reply:
@tztsai I think @vgro means a fundamentally simpler algorithm rather than a simpler implementation of this. Although having said that, if https://numpy.org/doc/stable/reference/generated/numpy.polynomial.polynomial.Polynomial.html is more efficient then there are lots of places we could use it.

Reply:
Well, huh. That's a reasonable improvement - interestingly it looks like the overhead in creating a `Polynomial` instance isn't negligible, but I don't think it's worth creating `Polynomial` instances in the underlying constants classes.

```python

In [25]: import timeit
In [26]: m1 = timeit.timeit(
    ...:     stmt="""po = 0.99983952 + (6.788260e-5) * tc
    ...: po += -(9.08659e-6) * tc * tc
    ...: po += (1.022130e-7) * tc * tc * tc
    ...: po += -(1.35439e-9) * tc * tc * tc * tc
    ...: po += (1.471150e-11) * tc * tc * tc * tc * tc
    ...: po += -(1.11663e-13) * tc * tc * tc * tc * tc * tc
    ...: po += (5.044070e-16) * tc * tc * tc * tc * tc * tc * tc
    ...: po += -(1.00659e-18) * tc * tc * tc * tc * tc * tc * tc * tc
    ...: """,
    ...:     setup="""import numpy as np
    ...: tc = np.array([10,20,30])
    ...: """,
    ...: )
In [27]: m2 = timeit.timeit(
    ...:     stmt="""pn = np.polynomial.Polynomial(cf)
    ...: po = pn(tc)
    ...: """,
    ...:     setup="""import numpy as np
    ...: tc = np.array([10,20,30])
    ...: cf = np.array([0.99983952 , 6.788260e-5 , -9.08659e-6 , 1.022130e-7 , -1.35439e-9 ,
    ...:                1.471150e-11, -1.11663e-13, 5.044070e-16, -1.00659e-18])
    ...: """,
    ...: )
In [28]: m3 = timeit.timeit(
    ...:     stmt="po = pn(tc)",
    ...:     setup="""import numpy as np
    ...: tc = np.array([10,20,30])
    ...: cf = np.array([0.99983952 , 6.788260e-5 , -9.08659e-6 , 1.022130e-7 , -1.35439e-9 ,
    ...:                1.471150e-11, -1.11663e-13, 5.044070e-16, -1.00659e-18])
    ...: pn = np.polynomial.Polynomial(cf)
    ...: """,
    ...: )
In [29]: print(m1, m2, m3)
40.16773283301154 29.03079683400574 23.08425579199684
```